router/module.go

```diff
@@ -225,117 +225,162 @@ func (am AppModule) OnChanCloseConfirm(ctx sdk.Context, portID, channelID string
+	if err != nil || m.Forward == nil {
+		// not a packet that should be forwarded
+		am.keeper.Logger(ctx).Debug("packetForwardMiddleware OnRecvPacket forward metadata does not exist")
+		return am.app.OnRecvPacket(ctx, packet, relayer)
+	}
+
+	metadata := m.Forward
+
+	if err := metadata.Validate(); err != nil {
 		return channeltypes.NewErrorAcknowledgement(err.Error())
 	}
 
-	if !parsedReceiver.ShouldForward {
-		return am.app.OnRecvPacket(ctx, packet, relayer)
+	ack := am.app.OnRecvPacket(ctx, packet, relayer)
+	if ack == nil || !ack.Success() {
+		return ack
 	}
 
-	// Modify packet data to process packet transfer for this chain, omitting forwarding info
-	newData := data
-	newData.Receiver = parsedReceiver.HostAccAddr
-	bz, err := transfertypes.ModuleCdc.MarshalJSON(&newData)
-	if err != nil {
-		return channeltypes.NewErrorAcknowledgement(err.Error())
+	denomOnThisChain := GetDenomForThisChain(
+		packet.DestinationPort, packet.DestinationChannel,
+		packet.SourcePort, packet.SourceChannel,
+		data.Denom,
+	)
+
+	amountInt, ok := sdk.NewIntFromString(data.Amount)
+	if !ok {
+		return channeltypes.NewErrorAcknowledgement(fmt.Sprintf("error parsing amount for forward: %s", data.Amount))
 	}
-	newPacket := packet
-	newPacket.Data = bz
-
-	ack := am.app.OnRecvPacket(ctx, newPacket, relayer)
-	if ack.Success() {
-		// recalculate denom, skip checks that were already done in app.OnRecvPacket
-		var err error
-		// TODO put denom handling in separate function
-		var denom string
-		if transfertypes.ReceiverChainIsSource(packet.GetSourcePort(), packet.GetSourceChannel(), newData.Denom) {
-			// remove prefix added by sender chain
-			voucherPrefix := transfertypes.GetDenomPrefix(packet.GetSourcePort(), packet.GetSourceChannel())
-			unprefixedDenom := newData.Denom[len(voucherPrefix):]
-
-			// coin denomination used in sending from the escrow address
-			denom = unprefixedDenom
-
-			// The denomination used to send the coins is either the native denom or the hash of the path
-			// if the denomination is not native.
-			denomTrace := transfertypes.ParseDenomTrace(unprefixedDenom)
-			if denomTrace.Path != "" {
-				denom = denomTrace.IBCDenom()
-			}
-		} else {
-			prefixedDenom := transfertypes.GetDenomPrefix(packet.GetDestPort(), packet.GetDestChannel()) + newData.Denom
-			denom = transfertypes.ParseDenomTrace(prefixedDenom).IBCDenom()
-		}
-		unit, err := sdk.ParseUint(newData.Amount)
-		if err != nil {
-			channeltypes.NewErrorAcknowledgement(err.Error())
-		}
-		var token = sdk.NewCoin(denom, sdk.NewIntFromUint64(unit.Uint64()))
 
-		var timeout time.Duration
-		if parsedReceiver.Timeout.Nanoseconds() > 0 {
-			timeout = parsedReceiver.Timeout
-		} else {
-			timeout = am.forwardTimeout
-		}
+	token := sdk.NewCoin(denomOnThisChain, amountInt)
 
-		var retries uint8
-		if parsedReceiver.ForwardRetries != nil {
-			retries = *parsedReceiver.ForwardRetries
-		} else {
-			retries = am.retriesOnTimeout
-		}
+	var timeout time.Duration
```

Comment:
nit: Could you name this timeoutDuration or relativeTimeout so it doesn't get confused with the absolute timeout set in the packet